repo/migrations/Migration029_test.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/OpenBazaar/openbazaar-go/pb"
+	"github.com/OpenBazaar/openbazaar-go/repo/migrations"
+	"github.com/OpenBazaar/openbazaar-go/schema"
+	"github.com/OpenBazaar/openbazaar-go/test/factory"
+	"io/ioutil"
+	"os"
+	"testing"
+)
+
+func TestMigration029(t *testing.T) {
+	var testRepo, err = schema.NewCustomSchemaManager(schema.SchemaContext{
+		DataPath:        schema.GenerateTempPath(),
+		TestModeEnabled: true,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err = testRepo.BuildSchemaDirectories(); err != nil {
+		t.Fatal(err)
+	}
+	defer testRepo.DestroySchemaDirectories()
+
+	var (
+		repoverPath         = testRepo.DataPathJoin("repover")
+		listingIndexPath    = testRepo.DataPathJoin("root", "listings.json")
+		testListingSlug     = "Migration029_test_listing"
+		testListingPath     = testRepo.DataPathJoin("root", "listings", testListingSlug+".json")
+		expectedListingHash = "QmfEr6qqLxRsjJhk1XPq2FBP6aiwG6w6Dwr1XepU1Rg1Wx"
+	)
+
+	listing := factory.NewListing(testListingSlug)
+	m := jsonpb.Marshaler{
+		Indent:       "    ",
+		EmitDefaults: true,
+	}
+
+	f, err := os.Create(testListingPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := m.Marshal(f, listing); err != nil {
+		t.Fatal(err)
+	}
+
+	index := []*migrations.Migration029_ListingData{extractListingData(listing)}
+	indexJSON, err := json.MarshalIndent(&index, "", "    ")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := ioutil.WriteFile(listingIndexPath, indexJSON, os.ModePerm); err != nil {
+		t.Fatal(err)
+	}
+
+	var migration migrations.Migration029
+	if err := migration.Up(testRepo.DataPath(), "", true); err != nil {
+		t.Fatal(err)
+	}
+
+	var listingIndex []migrations.Migration029_ListingData
+	listingsJSON, err := ioutil.ReadFile(listingIndexPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = json.Unmarshal(listingsJSON, &listingIndex); err != nil {
+		t.Fatal(err)
+	}
+
+	if listingIndex[0].Hash != expectedListingHash {
+		t.Errorf("Expected listing hash %s got %s", expectedListingHash, listingIndex[0].Hash)
+	}
+
+	assertCorrectRepoVer(t, repoverPath, "30")
+
+	if err := migration.Down(testRepo.DataPath(), "", true); err != nil {
+		t.Fatal(err)
+	}
+
+	assertCorrectRepoVer(t, repoverPath, "29")
+}
+
+func extractListingData(listing *pb.Listing) *migrations.Migration029_ListingData {
+	descriptionLength := len(listing.Item.Description)
+
+	contains := func(s []string, e string) bool {
+		for _, a := range s {
+			if a == e {
+				return true
+			}
+		}
+		return false
+	}
+
+	var shipsTo []string
+	var freeShipping []string
+	for _, shippingOption := range listing.ShippingOptions {
+		for _, region := range shippingOption.Regions {
+			if !contains(shipsTo, region.String()) {
+				shipsTo = append(shipsTo, region.String())
+			}
+			for _, service := range shippingOption.Services {
+				if service.Price == 0 && !contains(freeShipping, region.String()) {
+					freeShipping = append(freeShipping, region.String())
+				}
+			}
+		}
+	}
```

Comment:
What is all of this for? Is there something wrong with the factory data?

Reply:
We need to convert the `pb.Listing` that is generated by the factory to a `ListingData` to save in the index.

Reply:
Could we just change the Hash without any of the other struct needing to be defined? https://play.golang.org/p/yx8O1FDQHM1